realm/realm-library/src/objectServer/java/io/realm/SyncSession.java

```diff
@@ -114,5 +130,185 @@ protected void finalize() throws Throwable {
+                    );
+                }
+                return null;
+            }
+
+            @Override
+            protected void onSuccess(AuthenticateResponse response) {
+                RealmLog.debug("Session[%s]: Access token acquired", configuration.getPath());
+                if (!isClosed && !Thread.currentThread().isInterrupted()) {
+                    ObjectServerUser.AccessDescription desc = new ObjectServerUser.AccessDescription(
+                            response.getAccessToken(),
+                            configuration.getPath(),
+                            configuration.shouldDeleteRealmOnLogout()
+                    );
+                    getUser().getSyncUser().addRealm(configuration.getServerUrl(), desc);
+                    // schedule a token refresh before it expires
+                    if (nativeRefreshAccessToken(configuration.getPath(), getUser().getSyncUser().getAccessToken(configuration.getServerUrl()).value(), configuration.getServerUrl().toString())) {
+                        scheduleRefreshAccessToken(authServer, response.getAccessToken().expiresMs());
+
+                    } else {
+                        // token not applied, no refresh will be scheduled
+                        onGoingAccessTokenQuery.set(false);
+                    }
+                }
+            }
+
+            @Override
+            protected void onError(AuthenticateResponse response) {
+                onGoingAccessTokenQuery.set(false);
+                RealmLog.debug("Session[%s]: Failed to get access token (%d)", configuration.getPath(), response.getError().getErrorCode());
+                if (!isClosed && !Thread.currentThread().isInterrupted()) {
+                    errorHandler.onError(SyncSession.this, response.getError());
+                }
+            }
+        });
+        networkRequest = new RealmAsyncTaskImpl(task, SyncManager.NETWORK_POOL_EXECUTOR);
+    }
+
+    private void scheduleRefreshAccessToken(final AuthenticationServer authServer, long expireDateInMs) {
+            // calculate the delay time before which we should refresh the access_token,
+            // we adjust to 10 second to proactively refresh the access_token before the session
+            // hit the expire date on the token
+            long refreshAfter =  expireDateInMs - System.currentTimeMillis() - REFRESH_MARGIN_DELAY;
+            if (refreshAfter < 0) {
+                // Token already expired
+                RealmLog.debug("Expires time already reached for the access token, refresh soon");
+                // we avoid refreshing directly to avoid an edge case where the client clock is ahead
+                // of the server, causing all access_token received from the server to be always
+                // expired, we will flood the server with refresh token requests then, so adding
+                // a bit of delay is the best effort in this case.
+                refreshAfter = REFRESH_MARGIN_DELAY;
```

Comment:
There's a use case where the user can end-up flooding the ROS with queries.

If the user/device clock is ahead compared to the server, then the token you just received from the server will be considered always as expired & will issue immediately another refresh request. 10 seconds feels like a safety net for me to avoid flooding, but I'm open to suggestions